homeassistant/components/stream/worker.py

```diff
@@ -0,0 +1,142 @@
+"""Proides the worker thread needed for processing streams."""
+import asyncio
+from fractions import Fraction
+import io
+
+from .core import Segment, StreamBuffer
+
+
+def generate_audio_frame():
+    """Generate a blank audio frame."""
+    from av import AudioFrame
+    audio_frame = AudioFrame(format='dbl', layout='mono', samples=1024)
+    audio_bytes = b''.join(b'\x00\x00\x00\x00\x00\x00\x00\x00'
+                           for i in range(0, 1024))
+    audio_frame.planes[0].update(audio_bytes)
+    audio_frame.sample_rate = 44100
+    audio_frame.time_base = Fraction(1, 44100)
+    return audio_frame
+
+
+def create_stream_buffer(stream_output, video_stream, audio_frame):
+    """Create a new StreamBuffer."""
+    import av
+    a_packet = None
+    segment = io.BytesIO()
+    output = av.open(
+        segment, mode='w', format=stream_output.format)
+    vstream = output.add_stream(
+        stream_output.video_codec, video_stream.rate)
+    # Fix format
+    vstream.codec_context.format = \
+        video_stream.codec_context.format
+    # Check if audio is requested
+    astream = None
+    if stream_output.audio_codec:
+        astream = output.add_stream(
+            stream_output.audio_codec, 44100)
+        # Need to do it multiple times for some reason
+        while not a_packet:
+            a_packets = astream.encode(audio_frame)
+            if a_packets:
+                a_packet = a_packets[0]
+    return (a_packet, StreamBuffer(segment, output, vstream, astream))
+
+
+def stream_worker(hass, stream, quit_event):
+    """Handle consuming streams."""
+    import av
+    try:
+        video_stream = stream.container.streams.video[0]
+    except (KeyError, IndexError):
+        hass.getLogger().error("Stream has no video")
+        return
+
+    audio_frame = generate_audio_frame()
+
+    outputs = {}
+    first_packet = True
+    sequence = 1
+    audio_packets = {}
+
+    while not quit_event.is_set():
+        try:
+            packet = next(stream.container.demux(video_stream))
+            if packet.dts is None:
+                # If we get a "flushing" packet, the stream is done
+                raise StopIteration
+        except (av.AVError, StopIteration):
+            # End of stream, clear listeners and stop thread
+            for fmt, buffer in outputs.items():
+                asyncio.run_coroutine_threadsafe(
+                    stream.outputs[fmt].put(None), hass.loop)
```

Comment:
`put` should not be defined as coroutine, this line can be simplify

Reply:
asyncio Event is not thread safe.

Reply:
Following the code:

- class `Stream` has a dictionary self._outputs, returned as a property `self.stream`
- `stream.add_provider` will add a provider and set it on the stream
- `HlsStreamOutput()` is added as a provider to a stream inside `HlsPlaylistView`
- `HlsStreamOutput` extends `StreamOutput`, only adds properties itself
- `StreamOutput` defines `async def put(self, segment: Segment) -> None:`

I think that we can convert `put` to be a `callback`. And then we can use `hass.loop.call_soon_threadsafe`

Reply:
that works.  done.